src/index.ml

```diff
@@ -364,42 +399,76 @@ module Make (K : Key) (V : Value) (IO : IO) = struct
+          no_changes ()
+        else if force then (
+          Log.debug (fun l -> l "[%s] force sync" (Filename.basename t.root));
+          let index_path = index_path t.root in
+          if Sys.file_exists index_path then
+            let io =
+              IO.v ~fresh:false ~readonly:true ~generation ~fan_size:0L
+                index_path
+            in
+            let fan_out =
+              Fan.import ~hash_size:K.hash_size (IO.get_fanout io)
+            in
+            t.index <- Some { fan_out; io }
+          else () )
         else no_changes ()
 
-  let find t key =
-    let t = check_open t in
-    Log.info (fun l -> l "[%s] find %a" (Filename.basename t.root) K.pp key);
-    if t.config.readonly then sync_log t;
-    match t.log with
-    | None -> raise Not_found
-    | Some log -> (
+  let find_instance t key =
+    let find_if_exists ~name ~find = function
+      | None ->
+          Log.debug (fun l ->
+              l "[%s] %s is not present" (Filename.basename t.root) name);
+          raise Not_found
+      | Some e ->
+          let ans = find e key in
+          Log.debug (fun l ->
+              l "[%s] found in %s" (Filename.basename t.root) name);
+          ans
+    in
+    let find_log_index () =
+      try find_if_exists ~name:"log" ~find:(fun log -> Tbl.find log.mem) t.log
+      with Not_found -> (
+        match
+          find_if_exists ~name:"index" ~find:interpolation_search t.index
+        with
+        | Some e -> e
+        | None -> raise Not_found )
+    in
+    IO.Mutex.with_lock t.rename_lock (fun () ->
+        if t.config.readonly then sync_log t;
         try
-          let value = Tbl.find log.mem key in
-          Log.debug (fun l -> l "[%s] found in log" (Filename.basename t.root));
-          value
+          find_if_exists ~name:"log" ~find:(fun log -> Tbl.find log.mem) t.log
         with Not_found -> (
-          match t.index with
-          | None ->
-              Log.debug (fun l ->
-                  l "[%s] not found" (Filename.basename t.root));
-              raise Not_found
-          | Some index -> (
-              match interpolation_search index key with
-              | Some e ->
-                  Log.debug (fun l ->
-                      l "[%s] found in index" (Filename.basename t.root));
-                  e
-              | None ->
-                  Log.debug (fun l ->
-                      l "[%s] not found in index" (Filename.basename t.root));
-                  raise Not_found ) ) )
+          try
+            find_if_exists ~name:"log_async"
+              ~find:(fun log -> Tbl.find log.mem)
+              t.log_async
+          with Not_found -> (
+            match
+              find_if_exists ~name:"index" ~find:interpolation_search t.index
+            with
+            | Some e -> e
+            | None when t.config.readonly ->
+                sync_log ~force:true t;
+                find_log_index ()
```

Comment:
So in that case, if a read-only instance does not find the binding, it reads the whole `log` and re-imports the fanout again from a new index file descriptor, then retries to read, do I understand correctly?
Why is this necessary?

Reply:
The idea is that if the last steps of a merge interleave with a RO's sync_log, the RO instance does not see the latest update. 
 
> This bug was caused by something like* the following scenario:
- start with an empty log, for both RW and RO
> RW - writes value in log, then calls `merge`, the value is moved into index, but the generation is not set
> RO - calls `sync_log` and reads generation 

-`IO.get_generation`
> RW - clears log, sets generation
> RO - reads offset of log 

-`IO.force_offset`
> RO does not see any changes in neither generation nor offset. it does not update its index and therefore cannot find the value.

Reply:
But this situation can also appear if you start with a log of size 1 and add one entry in log_async during the merge. By the end of the merge you end up with a log of size 1, and RO could miss that there was an update.